src/qibolab/transpilers/general_connectivity.py

```diff
@@ -0,0 +1,366 @@
+
+        Args:
+            graph (networkx.Graph): current hardware qubit mapping.
+
+        Returns:
+            new_circuit (list): reduced circuit.
+        """
+        new_circuit = deepcopy(self._circuit_repr)
+        while new_circuit != [] and (new_circuit[0][0], new_circuit[0][1]) in graph.edges():
+            del new_circuit[0]
+        return new_circuit
+
+    def subgraph_init(self):
+        """Subgraph isomorphism initialization, NP-complete it can take a long time for large circuits."""
+        H = nx.Graph()
+        H.add_nodes_from([i for i in range(0, self._connectivity.number_of_nodes())])
+        GM = nx.algorithms.isomorphism.GraphMatcher(self._connectivity, H)
+        i = 0
+        H.add_edge(self._circuit_repr[i][0], self._circuit_repr[i][1])
+        while GM.subgraph_is_monomorphic() == True:
+            result = GM
+            i = i + 1
+            H.add_edge(self._circuit_repr[i][0], self._circuit_repr[i][1])
+            GM = nx.algorithms.isomorphism.GraphMatcher(self._connectivity, H)
+            if self._connectivity.number_of_edges() == H.number_of_edges() or i == len(self._circuit_repr) - 1:
+                G = nx.relabel_nodes(self._connectivity, result.mapping)
+                self._graph = G
+                self._mapping = result.mapping
+                return
+        G = nx.relabel_nodes(self._connectivity, result.mapping)
+        self._graph = G
+        self._mapping = result.mapping
+
+    def greedy_init(self):
+        """initialize the circuit with greedy algorithm let a maximum number of 2-qubit
+        gates can be applied without introducing any SWAP gate"""
+        nodes = self._connectivity.number_of_nodes()
+        keys = list(self._connectivity.nodes())
+        values = [i for i in range(nodes)]
+        final_mapping = {keys[i]: values[i] for i in range(len(keys))}
+        final_graph = nx.relabel_nodes(self._connectivity, final_mapping)
+        final_cost = len(self.reduce(final_graph))
+        for _ in range(self._init_samples):
+            random.shuffle(values)
+            mapping = {keys[i]: values[i] for i in range(len(keys))}
+            graph = nx.relabel_nodes(self._connectivity, mapping)
+            cost = len(self.reduce(graph))
+            if cost == 0:
+                self._graph = graph
+                self._mapping = mapping
+                return
+            if cost < final_cost:
+                final_graph = graph
+                final_mapping = mapping
+                final_cost = cost
+        self._graph = final_graph
+        self._mapping = final_mapping
+
+    def init_qubit_map(self):
+        """Initial circuit-hardware qubit mapping.
+
+        Returns:
+            qubit_map (np.array): array containing the initial qubit mapping of the circuit.
+        """
+        qubit_map = np.zeros((len(self._mapping.keys()),), dtype=int)
+        i = 0
+        for key in self._mapping.keys():
```

Comment:
```suggestion
        for key in self._mapping:
```

Iteration on a `dict` is by default on its keys